src/expr/src/interpret.rs

```diff
@@ -945,28 +988,250 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
+            _ => false,
+        }
+    }
+
+    const INTERESTING_VARIADIC_FUNCS: &[VariadicFunc] = {
+        use VariadicFunc::*;
+        &[Coalesce, Greatest, Least, And, Or]
+    };
+
+    fn variadic_typecheck(func: &VariadicFunc, args: &[ColumnType]) -> bool {
+        use VariadicFunc::*;
+        match func {
+            Coalesce | Greatest | Least => match args {
+                [] => true,
+                [first, rest @ ..] => rest
+                    .iter()
+                    .all(|t| first.scalar_type.base_eq(&t.scalar_type)),
+            },
+            And | Or => args
+                .iter()
+                .all(|t| t.scalar_type.base_eq(&ScalarType::Bool)),
+            _ => false,
+        }
+    }
+
+    fn gen_datums_for_type(typ: &ColumnType) -> BoxedStrategy<Datum<'static>> {
+        let mut values: Vec<Datum<'static>> = typ.scalar_type.interesting_datums().collect();
+        if typ.nullable {
+            values.push(Datum::Null)
+        }
+        select(values).boxed()
     }
 
     fn gen_column() -> impl Strategy<Value = (ColumnType, Datum<'static>)> {
-        any::<ColumnType>()
+        select(SCALAR_TYPES)
+            .prop_map(|t| t.nullable(true))
             .prop_filter("need at least one value", |c| {
                 c.scalar_type.interesting_datums().count() > 0
             })
             .prop_flat_map(|col| {
-                gen_datums_for_type(&col.scalar_type).prop_map(move |datum| (col.clone(), datum))
+                gen_datums_for_type(&col).prop_map(move |datum| (col.clone(), datum))
+            })
+    }
+
+    fn gen_expr_for_relation(
+        relation: &RelationType,
+    ) -> BoxedStrategy<(MirScalarExpr, ColumnType)> {
+        let column_gen = {
+            let column_types = relation.column_types.clone();
+            any::<Index>()
+                .prop_map(move |idx| {
+                    let id = idx.index(column_types.len());
+                    (MirScalarExpr::Column(id), column_types[id].clone())
+                })
+                .boxed()
+        };
+
+        let literal_gen = select(SCALAR_TYPES)
+            .prop_map(|s| s.nullable(true))
+            .prop_flat_map(|ct| {
+                let error_gen = any::<EvalError>().prop_map(Err).boxed();
+                let value_gen = gen_datums_for_type(&ct)
+                    .prop_map(move |datum| Ok(Row::pack_slice(&[datum])))
+                    .boxed();
+                error_gen.prop_union(value_gen).prop_map(move |result| {
+                    (MirScalarExpr::Literal(result, ct.clone()), ct.clone())
+                })
+            })
+            .boxed();
+
+        column_gen
+            .prop_union(literal_gen)
+            .prop_recursive(4, 64, 8, |self_gen| {
+                let unary_gen = (select(INTERESTING_UNARY_FUNCS), self_gen.clone())
+                    .prop_filter_map("unary func", |(func, (expr_in, type_in))| {
+                        if !unary_typecheck(&func, &type_in) {
```

Comment:
I wonder how much runtime overhead the "generate and reject"-based approach to producing well-typed terms will cause. In the future, I think that we can write some proptest functions that produce expressions of a given type and then use the expected types of the function to seed the expressions.

Reply:
Currently the overhead is significant but not overwhelming... this proptest takes under a second to complete.

I agree that it should be possible to write more typed generators. In practise one obstacle is the workings of `prop_recursive`, since that only allows a function to recurse on itself, where having variously-typed generators would be mutual recursion. The generate-and-reject approach allows me to use more of the builtin functionality here.